### PR TITLE
documents: prevent referenced document deletion

### PR DESCRIPTION
**Pull request by rerowep**
* Adds to the documents `reasons_not_to_delete` function the part of documents count.
* Adds `get_pids` to all `get_links_to_me` functions to get all attached pids or counts.

Co-Authored-by: Peter Weber <peter.weber@rero.ch>

## Why are you opening this PR?

- Which task/US does it implement?
- Which issue does it fix?

## Dependencies

My PR depends on the following `rero-ils-ui`'s PR(s):

* rero/rero-ils-ui#<xx>

## How to test?

- What command should I have to run to test your PR?
- What should I test through the UI?

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?
- [ ] Cypress tests successful?

### CODE REVIEW

**Review by zannkukai — APPROVED** [review]

File: rero_ils/modules/budgets/api.py [hunk 1]
```diff
@@ -67,16 +67,18 @@ class Budget(IlsRecord):
         }
     }
 
-    def get_number_of_acq_accounts(self):
+    def get_number_of_acq_accounts(self, get_pids=False):
```
Comment (zannkukai) [inline rero_ils/modules/budgets/api.py:70]: the method name suggests that we get the number of linked acq_account. Either change the signature, either don't play with `get_pids`

File: rero_ils/modules/budgets/api.py [hunk 2]
```diff
@@ -67,16 +67,18 @@ class Budget(IlsRecord):
...
+            return sorted_pids(query)
+        else:
+            return query.count()
 
-    def get_links_to_me(self):
+    def get_links_to_me(self, get_pids=False):
         """Get number of links."""
         links = {}
-        acq_accounts = self.get_number_of_acq_accounts()
+        acq_accounts = self.get_number_of_acq_accounts(get_pids=get_pids)
```
Comment (zannkukai) [inline rero_ils/modules/budgets/api.py:81]: Depending of first comment, adapt method call here

File: rero_ils/modules/circ_policies/api.py [hunk 3]
```diff
@@ -292,7 +292,7 @@ def reasons_to_keep(self):
             others['is_default'] = is_default
         return others
 
-    def get_links_to_me(self):
+    def get_links_to_me(self, get_pids=False):
         """Get number of links to policy."""
         links = {}
         return links
```
Comment (zannkukai) [inline rero_ils/modules/circ_policies/api.py:298]: Could you please place a `TODO` comment here to remember to have a better method implementation.

File: rero_ils/modules/acq_accounts/api.py [hunk 4]
```diff
@@ -76,16 +76,16 @@ def library_pid(self):
         """Shortcut for acq account library pid."""
         return extracted_data_from_ref(self.get('library'))
 
-    def get_number_of_acq_order_lines(self):
+    def get_number_of_acq_order_lines(self, get_pids=False):
```
Comment (zannkukai) [inline rero_ils/modules/acq_accounts/api.py:79]: please update method name or method signature.

File: rero_ils/modules/acq_order_lines/api.py [hunk 5]
```diff
@@ -136,11 +136,11 @@ def get_order(self):
         from ..acq_orders.api import AcqOrder
         return AcqOrder.get_record_by_pid(self.order_pid)
 
-    def get_number_of_acq_order_lines(self):
+    def get_number_of_acq_order_lines(self, get_pids=False):
```
Comment (zannkukai) [inline rero_ils/modules/acq_order_lines/api.py:139]: same remarks

File: rero_ils/modules/acq_orders/api.py [hunk 6]
```diff
@@ -112,11 +112,11 @@ def library_pid(self):
         """Shortcut for acquisition order library pid."""
         return extracted_data_from_ref(self.get('library'))
 
-    def get_number_of_acq_order_lines(self):
+    def get_number_of_acq_order_lines(self, get_pids=False):
```
Comment (zannkukai) [inline rero_ils/modules/acq_orders/api.py:115]: same

File: rero_ils/modules/documents/api.py [hunk 7]
```diff
@@ -114,54 +115,88 @@ def can_edit(self):
         # TODO: Make this condition on data
         return not self.harvested
 
-    def get_number_of_items(self):
+    def get_number_of_items(self, get_pids=False):
```
Comment (zannkukai) [inline rero_ils/modules/documents/api.py:118]: same

File: rero_ils/modules/documents/api.py [hunk 8]
```diff
@@ -114,54 +115,88 @@ def can_edit(self):
...
         """Get number of items for document."""
         from ..items.api import ItemsSearch
-        return ItemsSearch().filter(
-            'term', document__pid=self.pid).count()
+        query = ItemsSearch().filter('term', document__pid=self.pid)
+        return get_count_or_pids(query, get_pids=get_pids)
 
-    def get_number_of_holdings(self):
+    def get_number_of_holdings(self, get_pids=False):
```
Comment (zannkukai) [inline rero_ils/modules/documents/api.py:124]: same

File: rero_ils/modules/documents/api.py [hunk 9]
```diff
@@ -114,54 +115,88 @@ def can_edit(self):
...
         """Get number of holdings for document."""
         from ..holdings.api import HoldingsSearch
-        return HoldingsSearch().filter(
-            'term', document__pid=self.pid).count()
+        query = HoldingsSearch().filter('term', document__pid=self.pid)
+        return get_count_or_pids(query, get_pids=get_pids)
 
-    def get_number_of_loans(self):
+    def get_number_of_loans(self, get_pids=False):
```
Comment (zannkukai) [inline rero_ils/modules/documents/api.py:130]: same

File: rero_ils/modules/documents/api.py [hunk 10]
```diff
@@ -114,54 +115,88 @@ def can_edit(self):
...
             exclude_states=[
                 LoanState.CANCELLED,
                 LoanState.ITEM_RETURNED,
             ]
         )
-        return search.source().count()
+        return get_count_or_pids(query, get_pids=get_pids)
 
-    def get_number_of_acquisition_order_lines(self):
+    def get_number_of_acquisition_order_lines(self, get_pids=False):
```
Comment (zannkukai) [inline rero_ils/modules/documents/api.py:142]: same

File: rero_ils/modules/documents/api.py [hunk 11]
```diff
@@ -114,54 +115,88 @@ def can_edit(self):
...
         """Get number of acquisition order lines for document."""
-        return AcqOrderLinesSearch().filter(
-            'term', document__pid=self.pid).count()
+        query = AcqOrderLinesSearch().filter('term', document__pid=self.pid)
+        return get_count_or_pids(query, get_pids=get_pids)
+
+    def get_number_of_documents(self, get_pids=False):
```
Comment (zannkukai) [inline rero_ils/modules/documents/api.py:147]: same

File: rero_ils/modules/holdings/api.py [hunk 12]
```diff
@@ -370,13 +370,12 @@ def get_items(self):
                         item['call_number'] = issue_call_number
                     yield item
 
-    def get_number_of_items(self):
+    def get_number_of_items(self, get_pids=False):
```
Comment (zannkukai) [inline rero_ils/modules/holdings/api.py:373]: same

File: rero_ils/modules/item_types/api.py [hunk 13]
```diff
@@ -122,17 +122,18 @@ def exist_name_and_organisation_pid(cls, name, organisation_pid):
         except StopIteration:
             return None
 
-    def get_number_of_items(self):
+    def get_number_of_items(self, get_pids=False):
```
Comment (zannkukai) [inline rero_ils/modules/item_types/api.py:125]: same

File: rero_ils/modules/items/api/circulation.py [hunk 14]
```diff
@@ -1151,18 +1151,19 @@ def return_missing(self):
             LoanAction.RETURN_MISSING: None
         }
 
-    def get_number_of_loans(self):
+    def get_number_of_loans(self, get_pids=False):
```
Comment (zannkukai) [inline rero_ils/modules/items/api/circulation.py:1154]: same

File: rero_ils/modules/libraries/api.py [hunk 15]
```diff
@@ -293,27 +293,27 @@ def in_working_days(self, count, date=None):
             date = self.next_open(date=date)
         return date
 
-    def get_number_of_librarians(self):
+    def get_number_of_librarians(self, get_pids=False):
```
Comment (zannkukai) [inline rero_ils/modules/libraries/api.py:296]: same

File: rero_ils/modules/locations/api.py [hunk 16]
```diff
@@ -113,34 +113,33 @@ def get_library(self):
         library_pid = extracted_data_from_ref(self.get('library'))
         return Library.get_record_by_pid(library_pid)
 
-    def get_number_of_items(self):
+    def get_number_of_items(self, get_pids=False):
```
Comment (zannkukai) [inline rero_ils/modules/locations/api.py:116]: same

File: rero_ils/modules/locations/api.py [hunk 17]
```diff
@@ -113,34 +113,33 @@ def get_library(self):
...
         from ..items.api import ItemsSearch
-        results = ItemsSearch().filter(
-            'term', location__pid=self.pid).source().count()
-        return results
+        query = ItemsSearch() \
+            .filter('term', location__pid=self.pid)
+        return get_count_or_pids(query, get_pids=get_pids)
 
-    def get_number_of_loans(self):
+    def get_number_of_loans(self, get_pids=False):
```
Comment (zannkukai) [inline rero_ils/modules/locations/api.py:123]: same

File: rero_ils/modules/organisations/api.py [hunk 18]
```diff
@@ -151,15 +152,16 @@ def get_vendors(self):
         for pid in pids:
             yield Vendor.get_record_by_pid(pid)
 
-    def get_number_of_libraries(self):
+    def get_number_of_libraries(self, get_pids=False):
```
Comment (zannkukai) [inline rero_ils/modules/organisations/api.py:155]: same

File: rero_ils/modules/patron_transactions/api.py [hunk 19]
```diff
@@ -402,15 +402,17 @@ def events(self):
         for result in results:
             yield PatronTransactionEvent.get_record_by_pid(result.pid)
 
-    def get_number_of_patron_transaction_events(self):
+    def get_number_of_patron_transaction_events(self, get_pids=False):
```
Comment (zannkukai) [inline rero_ils/modules/patron_transactions/api.py:405]: same

File: rero_ils/modules/patron_types/api.py [hunk 20]
```diff
@@ -224,39 +224,37 @@ def get_linked_patron(self):
...
+    def get_number_of_patrons(self, get_pids=False):
         """Get number of patrons."""
-        return PatronsSearch()\
-            .filter('term', patron__type__pid=self.pid).source().count()
+        query = PatronsSearch()\
+            .filter('term', patron__type__pid=self.pid)
+        return get_count_or_pids(query, get_pids=get_pids)
 
-    def get_number_of_circ_policies(self):
+    def get_number_of_circ_policies(self, get_pids=False):
```
Comment (zannkukai) [inline rero_ils/modules/patron_types/api.py:233]: same

File: rero_ils/modules/patron_types/api.py [hunk 21]
```diff
@@ -224,39 +224,37 @@ def get_linked_patron(self):
         for result in results:
             yield Patron.get_record_by_pid(result.pid)
 
-    def get_number_of_patrons(self):
+    def get_number_of_patrons(self, get_pids=False):
```
Comment (zannkukai) [inline rero_ils/modules/patron_types/api.py:227]: same

File: rero_ils/modules/patrons/api.py [hunk 22]
```diff
@@ -472,23 +473,22 @@ def patron_type_pid(self):
         if self.get('patron', {}).get('type'):
             return extracted_data_from_ref(self.get('patron').get('type'))
 
-    def get_number_of_loans(self):
+    def get_number_of_loans(self, get_pids=False):
```
Comment (zannkukai) [inline rero_ils/modules/patrons/api.py:476]: same

File: rero_ils/modules/vendors/api.py [hunk 23]
```diff
@@ -77,19 +78,20 @@ def order_email(self):
             'order_contact', self.get('default_contact', {})
         ).get('email')
 
-    def count_links_to_me(self, search_class):
+    def count_links_to_me(self, search_class, get_pids=False):
```
Comment (zannkukai) [inline rero_ils/modules/vendors/api.py:81]: same

**Review by iGormilhit — CHANGES_REQUESTED** [review]
@rerowep I have issue to understand the commit message. We should talk about it. :slightly_smiling_face:

**Review by iGormilhit — CHANGES_REQUESTED** [review]
Commit message proposition:

```
documents: prevent referenced document deletion

* Verifies if a document is linked to another document before deleting
  it.
* Adds to the documents `reasons_not_to_delete` function a section with
  the documents count.
* Adds `get_pids` to all `get_links_to_me` functions to get all attached
  PIDs and counts.
```

**Review by iGormilhit — APPROVED** [review]

**Review by benerken — APPROVED** [review]

**Review by BadrAly — APPROVED** [review]

**Review by lauren-d — APPROVED** [review]

**Comment by lauren-d** [comment]
be careful github actions failed

**Pull request merged**